Let's import and clean the third sheet which is CustomerDemographic.

In [1]:
import numpy as np
import pandas as pd
import warnings
warnings.filterwarnings('ignore')

In [4]:
df3 = pd.read_excel('KPMG_VI_New_raw_data_update_final.xlsx', sheet_name='CustomerDemographic', header=1, engine='openpyxl')

In [5]:
df3.head()

customer_id      first_name  last_name  gender  \
0            1         Laraine  Medendorp       F   
1            2             Eli    Bockman    Male   
2            3           Arlin     Dearle    Male   
3            4          Talbot        NaN    Male   
4            5  Sheila-kathryn     Calton  Female   

   past_3_years_bike_related_purchases                  DOB  \
0                                   93  1953-10-12 00:00:00   
1                                   81  1980-12-16 00:00:00   
2                                   61  1954-01-20 00:00:00   
3                                   33  1961-10-03 00:00:00   
4                                   56  1977-05-13 00:00:00   

                job_title job_industry_category     wealth_segment  \
0     Executive Secretary                Health      Mass Customer   
1  Administrative Officer    Financial Services      Mass Customer   
2      Recruiting Manager              Property      Mass Customer   
3                     NaN                    IT      Mass Customer   
4           Senior Editor                   NaN  Affluent Customer   

  deceased_indicator                                            default  \
0                  N                                                 "'   
1                  N                       <script>alert('hi')</script>   
2                  N                                2018-02-01 00:00:00   
3                  N  () { _; } >_[$($())] { touch /tmp/blns.shellsh...   
4                  N                                                NIL   

  owns_car  tenure  
0      Yes    11.0  
1      Yes    16.0  
2      Yes    15.0  
3       No     7.0  
4      Yes     8.0

There's a column named default which has corrupted data points, let's drop that first.

In [6]:
del df3['default']

In [7]:
df3.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4000 entries, 0 to 3999
Data columns (total 12 columns):
 #   Column                               Non-Null Count  Dtype  
---  ------                               --------------  -----  
 0   customer_id                          4000 non-null   int64  
 1   first_name                           4000 non-null   object 
 2   last_name                            3875 non-null   object 
 3   gender                               4000 non-null   object 
 4   past_3_years_bike_related_purchases  4000 non-null   int64  
 5   DOB                                  3913 non-null   object 
 6   job_title                            3494 non-null   object 
 7   job_industry_category                3344 non-null   object 
 8   wealth_segment                       4000 non-null   object 
 9   deceased_indicator                   4000 non-null   object 
 10  owns_car                             4000 non-null   object 
 11  tenure                        

In [12]:
# check if all the columns are unique
cols = df3.columns
for col in cols:
    print(f'{col}\t --->{df3[col].is_unique}')

customer_id	 --->True
first_name	 --->False
last_name	 --->False
gender	 --->False
past_3_years_bike_related_purchases	 --->False
DOB	 --->False
job_title	 --->False
job_industry_category	 --->False
wealth_segment	 --->False
deceased_indicator	 --->False
owns_car	 --->False
tenure	 --->False


Now let's look at the missing data rules that i mentioned in TASK 1.1.

Remove the rows with missing values: One option is to remove the rows that contain missing values. However, this should be done with caution, as removing too many rows can significantly reduce the amount of data available for analysis. If the missing values are only a small percentage of the total dataset, it may be acceptable to remove them.

Impute missing values: Another option is to impute the missing values with estimated or predicted values. There are several methods for imputing missing values, such as mean or median imputation, mode imputation, and regression imputation. Mean or median imputation can be used for numeric columns like DOB, while mode imputation can be used for categorical columns like job_titles.

Create a separate category for missing values: Depending on the type of missing data, you may also choose to create a separate category for the missing values. This is typically done for categorical variables, where the missing value is coded as a separate category, such as "unknown" or "not applicable"

In [14]:
# missing values in %
df3.isnull().mean()*100

customer_id                             0.000
first_name                              0.000
last_name                               3.125
gender                                  0.000
past_3_years_bike_related_purchases     0.000
DOB                                     2.175
job_title                              12.650
job_industry_category                  16.400
wealth_segment                          0.000
deceased_indicator                      0.000
owns_car                                0.000
tenure                                  2.175
dtype: float64

ok, job_title and job_industry_category has the major missing data points. Here we can drop column with the most missing values by using the thresh= in dropna method and by default it is set to rows.

I'm only going to keep those columns that have minimum of 70% non-NaN values.

In [17]:
# Let's start imputing missing data. for most of the cases mode() is a better option for filling missing data.
mode_DOB = df3['DOB'].mode()[0]
df3['DOB'].fillna(mode_DOB, inplace=True)


In [18]:
df3['DOB'].isnull().sum()

0

job_title & job_industry_category both are missing about 12.65% & 16.4% of data respectively, using ffill or bfill to fill in missing values can be a useful approach in some cases, it is generally not recommended as the default method. Instead, it is best to use a method that is appropriate for the specific dataset and analysis.

here i'm going to fill in the missing values with the mode of the non-missing values in the column. This approach assumes that the mode is a reasonable estimate for the missing values and may be appropriate if the missing values are distributed similarly to the non-missing values.

In [19]:
# fillna for job_title
mode_job_title = df3['job_title'].mode()[0]
df3['job_title'].fillna(mode_job_title, inplace=True)

In [20]:
df3['job_title'].isnull().sum()

0

In [21]:
# fillna for job_industry_category
mode_job_indus = df3['job_industry_category'].mode()[0]
df3['job_industry_category'].fillna(mode_job_indus, inplace=True)

In [22]:
df3['job_industry_category'].isnull().sum()

0

In [23]:
df3['tenure'].value_counts()

tenure
7.0     235
5.0     228
11.0    221
10.0    218
16.0    215
8.0     211
18.0    208
12.0    202
9.0     200
14.0    200
6.0     192
13.0    191
4.0     191
17.0    182
15.0    179
1.0     166
3.0     160
19.0    159
2.0     150
20.0     96
22.0     55
21.0     54
Name: count, dtype: int64

In [24]:
# for tenure we're going to fill missing with a mean or median
mean_tenure = round(df3['tenure'].mean())
df3['tenure'].fillna(mean_tenure, inplace=True)

In [25]:
df3['tenure'].isnull().sum()

0

In [26]:
# Let's format the data
df3['gender'].value_counts()

gender
Female    2037
Male      1872
U           88
F            1
Femal        1
M            1
Name: count, dtype: int64

In [27]:
name_format = {'U': 'Unidentified', 'Femal': 'Female', 'M': 'Male', 'F': 'Female'}
df3['gender'] = df3['gender'].replace(name_format)

In [28]:
df3.isnull().sum()

customer_id                              0
first_name                               0
last_name                              125
gender                                   0
past_3_years_bike_related_purchases      0
DOB                                      0
job_title                                0
job_industry_category                    0
wealth_segment                           0
deceased_indicator                       0
owns_car                                 0
tenure                                   0
dtype: int64

Let's set standard Boolean values for columns deceased_indicator, owns_car

In [30]:
df3['deceased_indicator'].value_counts()

deceased_indicator
N    3998
Y       2
Name: count, dtype: int64

In [31]:
# deceased_indicator
df3['deceased_indicator'].replace({'N': False, 'Y':True},inplace=True)

In [32]:
df3['owns_car'].value_counts()

owns_car
Yes    2024
No     1976
Name: count, dtype: int64

In [35]:
# owns_car
df3['owns_car'].replace({'Yes': True, 'No': False}, inplace=True)

In [36]:
df3

customer_id      first_name  last_name        gender  \
0               1         Laraine  Medendorp        Female   
1               2             Eli    Bockman          Male   
2               3           Arlin     Dearle          Male   
3               4          Talbot        NaN          Male   
4               5  Sheila-kathryn     Calton        Female   
...           ...             ...        ...           ...   
3995         3996         Rosalia   Halgarth        Female   
3996         3997          Blanch     Nisuis        Female   
3997         3998          Sarene    Woolley  Unidentified   
3998         3999       Patrizius        NaN          Male   
3999         4000           Kippy    Oldland          Male   

      past_3_years_bike_related_purchases                  DOB  \
0                                      93  1953-10-12 00:00:00   
1                                      81  1980-12-16 00:00:00   
2                                      61  1954-01-20 00:00:00   
3                                      33  1961-10-03 00:00:00   
4                                      56  1977-05-13 00:00:00   
...                                   ...                  ...   
3995                                    8  1975-08-09 00:00:00   
3996                                   87  2001-07-13 00:00:00   
3997                                   60  1978-01-30 00:00:00   
3998                                   11  1973-10-24 00:00:00   
3999                                   76  1991-11-05 00:00:00   

                                 job_title job_industry_category  \
0                      Executive Secretary                Health   
1                   Administrative Officer    Financial Services   
2                       Recruiting Manager              Property   
3     Business Systems Development Analyst                    IT   
4                            Senior Editor         Manufacturing   
...                                    ...                   ...   
3995                 VP Product Management                Health   
3996                       Statistician II         Manufacturing   
3997                     Assistant Manager                    IT   
3998  Business Systems Development Analyst         Manufacturing   
3999                  Software Engineer IV         Manufacturing   

         wealth_segment  deceased_indicator  owns_car  tenure  
0         Mass Customer               False      True    11.0  
1         Mass Customer               False      True    16.0  
2         Mass Customer               False      True    15.0  
3         Mass Customer               False     False     7.0  
4     Affluent Customer               False      True     8.0  
...                 ...                 ...       ...     ...  
3995      Mass Customer               False     False    19.0  
3996     High Net Worth               False      True     1.0  
3997     High Net Worth               False     False    11.0  
3998  Affluent Customer               False      True    10.0  
3999  Affluent Customer               False     False    11.0  

[4000 rows x 12 columns]

In [38]:
df3.to_csv('Task 1.3-CustomerDemographic.csv', index=False)